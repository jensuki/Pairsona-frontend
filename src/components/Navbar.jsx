import { useContext } from 'react';
import { NavLink, useNavigate } from 'react-router-dom';
import UserContext from '../context/UserContext';
import Api from '../helpers/api';
import pairsonaLogo from '../assets/logos/pairsona_logo.png';
import '../styles/Navbar.css';

/** navbar component
 *
 * displays nav links for pairona
 * conditionally displayed links based on currUser stte
 * show notifications for requests and unread messages for logged in user
*/

const Navbar = () => {
    const {
        currUser,
        logout,
        hasNotifications,
        hasNewMessages,
    } = useContext(UserContext);
    const navigate = useNavigate();

    // handle logo click
    const handleLogoClick = (e) => {
        e.preventDefault();
        if (!currUser || Api.isTokenExpired()) {
            logout();
            navigate('/');
        } else {
            navigate('/');
        }
    };

    return (
        <nav className="Navbar">

            {/* brand logo */}
            <div className="Navbar-brand">
                <a href="/" onClick={handleLogoClick} aria-label="Home">
                    <img src={pairsonaLogo} className="Navbar-logo" alt="Pairsona-Logo" />
                </a>
            </div>

            {/* nav links */}
            {currUser ? (
                <ul className="Navbar-links">
                    {/* link to curr user's profile */}
                    <li className="nav-item">
                        <NavLink
                            to="/profile"
                            aria-label="View Profile"
                            // apply 'active' class only when viewing curr users profile
                            className={({ isActive }) =>
                                isActive && location.pathname === '/profile' ? 'active' : ''}
                        >
                            {currUser.firstName}
                        </NavLink>
                    </li>

                    {/* link to find matches */}
                    <li className="nav-item">
                        <NavLink
                            to="/matches"
                            aria-label="Find Matches">
                            Find Matches
                        </NavLink>
                    </li>

                    {/* notification link with indicator */}
                    <li className="nav-item">
                        <NavLink to="/requests" aria-label="View Notifications">
                            Notifications
                            {hasNotifications && <span className="notification-dot"></span>}
                        </NavLink>
                    </li>

                    {/* connections link with unread message indicator */}
                    <li className="nav-item">
                        <NavLink to="/connections" aria-label="View Connections">
                            Connections
                            {hasNewMessages.length > 0 && <span className="notification-dot"></span>}
                        </NavLink>
                    </li>

                    {/* logout button */}
                    <li className="nav-item">
                        <button onClick={logout} className="logout-btn">
                            Log Out
                        </button>
                    </li>
                </ul>
            ) : (
                <ul className="Navbar-links">
                    {/* public links for users not logged in */}
                    <li className="nav-item">
                        <NavLink to="/login">Login</NavLink>
                    </li>
                    <li className="nav-item">
                        <NavLink to="/signup">Sign Up</NavLink>
                    </li>
                </ul>
            )}
        </nav>
    )
}

export default Navbar;