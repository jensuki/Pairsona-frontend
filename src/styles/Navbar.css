.Navbar {
    display: flex;
    max-width: 100%;
    justify-content: space-between;
    align-items: center;
    padding: 0;
    background-color: #f4f4f4;
}

.Navbar-brand img {
    margin: 10px;
    width: 40px;
    height: 26px;
}

.Navbar-links {
    list-style: none;
    display: flex;
    gap: .7rem;
}

.Navbar-links a {
    text-decoration: none;
    color: #211f1f;
    font-weight: 450;
    transition: color 0.2s ease;

    &:hover {
        color: var(--primary-green);
    }
}

.Navbar-links a.active {
    font-weight: 550;
    color: var(--primary-blue);
}

.Navbar .logout-btn {
    font-family: inherit;
    background: none;
    cursor: pointer;
    color: black;
    padding: 0;
}

.nav-item {
    position: relative;
    margin-right: 12px;
}

.notification-dot {
    position: absolute;
    top: 8px;
    right: -10px;
    width: 8px;
    height: 8px;
    background-color: rgb(247, 76, 76);
    border-radius: 50%;
}