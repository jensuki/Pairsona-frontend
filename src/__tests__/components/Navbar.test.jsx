import { render, screen, fireEvent } from '@testing-library/react';
import { UserProvider } from '../../test_common/setupTests';
import { demoUser } from '../../test_common/mockData';
import { MemoryRouter } from 'react-router-dom';
import Navbar from '../../components/Navbar';
import { expect, describe, test, vi } from 'vitest';

describe('Navbar component', () => {
    test('renders without crashing', () => {
        render(
            <MemoryRouter>
                <UserProvider>
                    <Navbar />
                </UserProvider>
            </MemoryRouter>
        )

        expect(screen.getByRole('navigation')).toBeInTheDocument();
    });

    test('matches snapshot for anon user', () => {
        const { asFragment } = render(
            <MemoryRouter>
                <UserProvider>
                    <Navbar />
                </UserProvider>
            </MemoryRouter>
        )

        expect(asFragment()).toMatchSnapshot();
    });

    test('matches snapshot for logged in user', () => {
        const { asFragment } = render(
            <MemoryRouter>
                <UserProvider currUser={demoUser}>
                    <Navbar />
                </UserProvider>
            </MemoryRouter>
        )

        expect(asFragment()).toMatchSnapshot();
    });

    test('renders proper links for anon user', () => {
        const { getByText } = render(
            <MemoryRouter>
                <UserProvider currUser={null}>
                    <Navbar />
                </UserProvider>
            </MemoryRouter>
        )

        expect(getByText('Login')).toBeInTheDocument();
        expect(getByText('Sign Up')).toBeInTheDocument();
    })

    test('renders proper links for logged in user', () => {
        render(
            <MemoryRouter>
                <UserProvider currUser={demoUser}>
                    <Navbar />
                </UserProvider>
            </MemoryRouter>
        )

        expect(screen.getByText(demoUser.firstName)).toBeInTheDocument();
        expect(screen.getByText('Find Matches')).toBeInTheDocument();
        expect(screen.getByText('Notifications')).toBeInTheDocument();
        expect(screen.getByText('Connections')).toBeInTheDocument();
        expect(screen.getByText('Log Out')).toBeInTheDocument();
    })

    test('shows notification indicator for logged in user', () => {
        render(
            <MemoryRouter>
                <UserProvider currUser={demoUser} hasNotifications={true}>
                    <Navbar />
                </UserProvider>
            </MemoryRouter>
        )
        // confirm the notification dot is rendered
        const notificationDot = screen.getByText('Notifications').querySelector('.notification-dot')
        expect(notificationDot).toBeInTheDocument();
    })

    test('shows unread message indicator for logged in user', () => {
        const mockUnreadMsg = ['msg1'];

        render(
            <MemoryRouter>
                <UserProvider currUser={demoUser} hasNewMessages={mockUnreadMsg}>
                    <Navbar />
                </UserProvider>
            </MemoryRouter>
        )
        // confirm the notification dot is rendered
        const unreadMsgDot = screen.getByText('Connections').querySelector('.notification-dot');
        expect(unreadMsgDot).toBeInTheDocument();
    })

    test('calls logout func when log out button is clicked', () => {
        const mockLogout = vi.fn();

        render(
            <MemoryRouter>
                <UserProvider currUser={demoUser} logout={mockLogout}>
                    <Navbar />
                </UserProvider>
            </MemoryRouter>
        )

        fireEvent.click(screen.getByText('Log Out'));
        expect(mockLogout).toHaveBeenCalled();
    })
})
