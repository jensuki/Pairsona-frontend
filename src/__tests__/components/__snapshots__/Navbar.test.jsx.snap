// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Navbar component > matches snapshot for anon user 1`] = `
<DocumentFragment>
  <nav
    class="Navbar"
  >
    <div
      class="Navbar-brand"
    >
      <a
        aria-label="Home"
        href="/"
      >
        <img
          alt="Pairsona-Logo"
          class="Navbar-logo"
          src="/src/assets/logos/pairsona_logo.png"
        />
      </a>
    </div>
    <ul
      class="Navbar-links"
    >
      <li
        class="nav-item"
      >
        <a
          class=""
          data-discover="true"
          href="/login"
        >
          Login
        </a>
      </li>
      <li
        class="nav-item"
      >
        <a
          class=""
          data-discover="true"
          href="/signup"
        >
          Sign Up
        </a>
      </li>
    </ul>
  </nav>
</DocumentFragment>
`;

exports[`Navbar component > matches snapshot for logged in user 1`] = `
<DocumentFragment>
  <nav
    class="Navbar"
  >
    <div
      class="Navbar-brand"
    >
      <a
        aria-label="Home"
        href="/"
      >
        <img
          alt="Pairsona-Logo"
          class="Navbar-logo"
          src="/src/assets/logos/pairsona_logo.png"
        />
      </a>
    </div>
    <ul
      class="Navbar-links"
    >
      <li
        class="nav-item"
      >
        <a
          aria-label="View Profile"
          class=""
          data-discover="true"
          href="/profile"
        >
          Test
        </a>
      </li>
      <li
        class="nav-item"
      >
        <a
          aria-label="Find Matches"
          class=""
          data-discover="true"
          href="/matches"
        >
          Find Matches
        </a>
      </li>
      <li
        class="nav-item"
      >
        <a
          aria-label="View Notifications"
          class=""
          data-discover="true"
          href="/requests"
        >
          Notifications
        </a>
      </li>
      <li
        class="nav-item"
      >
        <a
          aria-label="View Connections"
          class=""
          data-discover="true"
          href="/connections"
        >
          Connections
        </a>
      </li>
      <li
        class="nav-item"
      >
        <button
          class="logout-btn"
        >
          Log Out
        </button>
      </li>
    </ul>
  </nav>
</DocumentFragment>
`;
